### ISSUE TITLE
Optimistic Offline Lock

### ISSUE THREAD

**Issue by bricelam** [aspnet/EntityFramework#2195]
Many databases (including PostgreSQL and SQLite) don't have an equivalent to SQL Server's `rowversion`. We could improve our support for optimistic concurrency on those stores by implementing the [Optimistic Offline Lock](http://www.martinfowler.com/eaaCatalog/optimisticOfflineLock.html) pattern in the framework.

This would be similar to having a `rowversion` property in the entity except that the framework would manage the value. When the entity is saved, somewhere in the update pipeline, we would automatically increment the properties value. The DML would look something like this:
```SQL
UPDATE Person
SET Name = "Brice", RowVersion = RowVersion + 1
WHERE RowVersion = 23;
```

**Comment by anpete** [comment]
+1

It's also common to have a convention for this. I.e. Look for a property called "Version", "LockVersion" etc.

**Comment by divega** [comment]
Not sure it is a constrain but we don't need to express RowVersion = RowVersoin + 1 in the update pipeline, just passing the value 24 for the example above would work, correct?

**Comment by anpete** [comment]
Yes, it just means we need to be able to do the client side increment for all potential types.

**Comment by roji** [comment]
+1, needed for PostgreSQL which has no rowversion (although a workaround with generating values on update via triggers is probably possible)

**Comment by pantonis** [comment]
+1, We need it :) Not all databases are so productive working with them like sql server. PostgreSQL like roji mentioned does not support rowversion.

**Comment by roji** [comment]
FYI I found out PostgreSQL does have an automatically-updated xmin column which could be used as a concurrency token, much like SqlServer's rowversion, so this may be less important for PostgreSQL. See https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/issues/19.

**Comment by bubibubi** [comment]
+1
If you implement this please keep in mind that properties marked as Timestamp are array of bytes.
Changing the value could be not only an increment.

**Comment by anpete** [comment]
@bricelam I [implemented this](https://github.com/aspnet/EntityFrameworkCore/blob/dev/samples/OracleProvider/src/OracleProvider/Migrations/OracleMigrationsSqlGenerator.cs#L795) in the Oracle sample provider by auto-generating triggers in Migrations when creating row version columns.

**Comment by m4rs2k** [comment]
Is there any vote system to increase the priority of this feature?

Thx

**Comment by ajcvickers** [comment]
@m4rs2k Yes, 👍 this issue.

**Comment by asbjornu** [comment]
As reported in #32083, I'm very eager to see this feature implemented. Beyond waiting for it to be implemented in EF Core (which looks like it's going to take a while), is it possible to implement through some custom code? If so, how would that look like?

**Comment by roji** [comment]
@asbjornu you can override SaveChanges and have code there which sets the value of the concurrency token to an incremented value. Or if you want this to happen in the database, you can set up triggers to do that there.

**Comment by roji** [comment]
@ComptonAlvaro first, to ensure your business rule of no more than 5 items per order, I'd recommend considering an insert trigger. That would allow you to ensure, at the database level, that your invariant is always preserved.

Second, it seems you're asking for how to manage a concurrency token - that doesn't seem really related to the 5-items-per-order rule above. In any case, yes - you can have a shadow property that's configured as a concurrency token with EF, and then override SaveChanges to enumerate over all modified entities in the context and update the concurrency token.
